exec/java-exec/src/main/java/org/apache/drill/exec/work/WorkManager.java

```diff
@@ -433,6 +442,22 @@ public void run() {
     public StatusThread() {
       // assume this thread is created by a non-daemon thread
       setName("WorkManager.StatusThread");
+  }
+  private void pollShutdown(Drillbit drillbit) throws IOException, InterruptedException {
+      final Path path = FileSystems.getDefault().getPath(System.getenv("DRILL_PID_DIR"));
+      try (final WatchService watchService = FileSystems.getDefault().newWatchService()) {
+        path.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_CREATE);
+        while (true) {
+          final WatchKey wk = watchService.take();
```

Comment:
This design is not correct and has issues.

- You are using `WorkManager` `StatusThread` for registering a `WatchService` and then doing a blocking call `take()` which will make this thread stuck until there is any event seen by `WatchService`. Hence `StatusThread` will not be able to perform it's actual job of sending status of `runningFragments` to it's `Foreman`.
- The `pollShutdown` method is calling `Drillbit.close()` which is in `StatusThread`. Now `Drillbit.close() `calls close on `WorkManager` and which calls `interrupt` on `StatusThread`. There is a cycle here. What it means to call interrupt on itself ? Usually interrupt is used by another thread to wake up a blocking thread. I think based on interrupt call the interrupt flag of status thread will be set. Then when it comes out of `pollShutdown` and calls sleep it will hit interrupted exception. It might work but I don't like the idea where WorkManager which is part of Drillbit has a thread which calling `Drillbit.close()`. 

Please create a separate thread within Drillbit class itself to do this. But in `Drillbit::close()` method you need to still `interrupt` this new thread for cases when `close` it called from some other path. But you can maintain a flag which will indicate if `Thread.interrupt()` should be called or not.

Reply:
@sohami I have made the changes based on your comments. Please let me know if you have any comments.